File: src/Vellum/web/js/semantic/saved-research.js
// Saved research: one Research (semantic/research.js) or Collection research (semantic/collection-research.js)
// result, written down so it can be read again without asking the question a second time. Kept by the host
// beside the recent list and the collections (Services/SavedResearch.cs, %LOCALAPPDATA%\Vellum\research.json).
//
//   { v, savedAt, question, source, document, collection, summary, sufficient, evidence: [...] }
//
//   v           the record's version, so an older file can still be read
//   savedAt     when it was kept, ISO 8601
//   question    what was asked, as it was typed
//   source      which research answered it: a value of provenance.js SOURCES
//   document    the document it was asked of (documentRef), for document research; null otherwise
//   collection  { id, name } the collection it was asked of, for collection research; null otherwise
//   summary     the one line Vellum wrote from the matches at the time — kept as it was, never rewritten
//   evidence    the ranked passages as they were found: { id, kind, page, text, matched, box, name, path,
//               provenance } — each one's provenance record exactly as semantic/provenance.js made it
//
// What this is not: a cache that refreshes, a second search engine, a second document index or a document
// identity. Opening a saved result never reads a PDF, never searches and never recomputes anything — what was
// found then is what is shown, evidence and summary alike. The documents are referenced the way the evidence
// already referenced them (a path, and the content key of the bytes that were read); a file that has moved,
// changed or gone is a file this result can no longer open, and it says so rather than looking for another.
//
// Pure: no reading, no writing, no network, no state.

import { PROVENANCE_VERSION, SOURCES } from './provenance.js';

export { SOURCES };

export const SAVED_RESEARCH_VERSION = 1;
export const MAX_NAME_LENGTH = 80;

const text = (value) => (typeof value === 'string' && value ? value : null);
const box4 = (value) => (Array.isArray(value) && value.length === 4 && value.every(Number.isFinite) ? Object.freeze([...value]) : null);
const page = (value) => (Number.isInteger(value) && value > 0 ? value : null);
const terms = (value) => Object.freeze((Array.isArray(value) ? value : []).filter((t) => typeof t === 'string'));

/** A document reference as saved: what provenance.js already froze, read back with nothing guessed. */
function docRef(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const contentKey = text(raw.contentKey);
  return Object.freeze({
    name: text(raw.name),
    path: text(raw.path),
    contentKey,
    reference: raw.reference === 'content' || raw.reference === 'path' ? raw.reference : (contentKey ? 'content' : 'path'),
  });
}

/** A provenance record as saved. Every field is kept as it stands; a damaged one is dropped, not repaired. */
function savedProvenance(raw) {
  if (!raw || typeof raw !== 'object') return null;
  const object = raw.object && typeof raw.object === 'object' ? raw.object : {};
  const runIds = Array.isArray(object.runIds) ? Object.freeze(object.runIds.filter((id) => typeof id === 'string')) : null;
  return Object.freeze({
    v: Number.isInteger(raw.v) ? raw.v : PROVENANCE_VERSION,
    source: text(raw.source) ?? SOURCES.document,
    document: docRef(raw.document) ?? docRef({}),
    page: page(raw.page),
    box: box4(raw.box),
    object: Object.freeze({ id: text(object.id), kind: text(object.kind), blockId: text(object.blockId), runIds }),
  });
}

/** One piece of evidence as saved: the passage, where it is, and where it came from. */
function savedEvidence(item) {
  if (!item || typeof item !== 'object') return null;
  const provenance = savedProvenance(item.provenance);
  const number = page(item.number) ?? page(item.page) ?? provenance?.page ?? null;
  const quote = typeof item.text === 'string' ? item.text : null;
  if (quote === null || number === null) return null;
  return Object.freeze({
    id: text(item.id),
    kind: text(item.kind),
    page: number,
    text: quote,
    matched: terms(item.matched),
    box: box4(item.box) ?? provenance?.box ?? null,
    // The file the passage came from, as the research named it: a collection item carries its own, one
    // document takes the document the research was asked of.
    name: text(item.name) ?? provenance?.document?.name ?? null,
    path: text(item.path) ?? provenance?.document?.path ?? null,
    provenance,
  });
}

/**
 * A result made ready to keep. Takes what Research and Collection research already return:
 *
 *   question    the question as it was asked
 *   source      a value of SOURCES; document research by default
 *   document    the documentRef the research was asked of (document research)
 *   collection  { id, name } the research was asked of (collection research)
 *   summary     Vellum's line about the matches, as it was shown
 *   sufficient  whether there was enough evidence
 *   evidence    the ranked passages, each with its provenance
 *   savedAt     when it is being kept (now by default)
 *
 * Frozen and JSON-safe: what this holds is exactly what the host writes and what opening it gives back.
 */
export function savedResearchRecord({
  question = '', source = SOURCES.document, document = null, collection = null,
  summary = '', sufficient = false, evidence = [], savedAt = new Date().toISOString(),
} = {}) {
  const items = (Array.isArray(evidence) ? evidence : []).map(savedEvidence).filter(Boolean);
  return Object.freeze({
    v: SAVED_RESEARCH_VERSION,
    savedAt: text(savedAt) ?? new Date().toISOString(),
    question: String(question ?? ''),
    source: text(source) ?? SOURCES.document,
    document: docRef(document),
    collection: collection && typeof collection === 'object'
      ? Object.freeze({ id: text(collection.id), name: text(collection.name) })
      : null,
    summary: String(summary ?? ''),
    sufficient: Boolean(sufficient) && items.length > 0,
    evidence: Object.freeze(items),
  });
}

/**
 * A stored result read back, or null when it is damaged beyond reading (not an object, or from a version this
 * Vellum doesn't know). Passages that can't be read are dropped; the rest of the result still opens.
 */
export function readSavedResearch(raw) {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return null;
  if (Number.isInteger(raw.v) && raw.v > SAVED_RESEARCH_VERSION) return null;
  return savedResearchRecord(raw);
}

/** Every document the evidence quotes, once, in the order it is first quoted: what the host is told to watch. */
export function savedResearchPaths(record) {
  const paths = [];
  for (const item of record?.evidence ?? []) {
    const path = item.path;
    if (path && !paths.some((p) => p.toLowerCase() === path.toLowerCase())) paths.push(path);
  }
  return paths;
}

/** What the result was asked of, in one line: the collection's name, or the document's. */
export function savedResearchSubject(record) {
  if (record?.collection?.name) return record.collection.name;
  return record?.document?.name ?? record?.evidence?.[0]?.name ?? 'this document';
}

/**
 * The saved evidence with each passage told whether its document is still there. `documents` is what the host
 * listed for the item ([{ path, exists }]); a document it doesn't name is treated as missing, because the host
 * lists exactly the files the result quotes. Nothing is searched for and no path is guessed at: a missing
 * document stays the one the evidence names, marked `missing`, and can't be opened.
 */
export function withAvailability(record, documents = []) {
  const known = new Map((Array.isArray(documents) ? documents : [])
    .filter((d) => d && typeof d.path === 'string')
    .map((d) => [d.path.toLowerCase(), Boolean(d.exists)]));
  const evidence = (record?.evidence ?? []).map((item) => Object.freeze({
    ...item,
    missing: !(item.path && known.get(item.path.toLowerCase()) === true),
  }));
  return Object.freeze({ ...record, evidence: Object.freeze(evidence), missing: evidence.filter((e) => e.missing).length });
}

/** A name for a saved result, cleaned the way the host cleans it; empty when there is nothing usable. */
export function cleanSavedName(name) {
  return String(name ?? '').split(/\s+/).filter(Boolean).join(' ').slice(0, MAX_NAME_LENGTH);
}

/** The name Vellum offers when a result is saved: the question, shortened, else what it was asked of. */
export function suggestedName(record) {
  return cleanSavedName(record?.question) || cleanSavedName(savedResearchSubject(record)) || 'Saved research';
}

File: src/Vellum/Services/SavedResearch.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace Vellum.Services;

public sealed class SavedResearchItem
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public DateTimeOffset CreatedAt { get; set; }
    /// <summary>Full paths of the documents the evidence was quoted from. The files are never copied or opened here.</summary>
    public List<string> Paths { get; set; } = [];
    /// <summary>The result exactly as the page saved it (semantic/saved-research.js). The host never reads inside it.</summary>
    public JsonNode? Result { get; set; }
}

/// <summary>
/// Research results a person chose to keep, in %LOCALAPPDATA%\Vellum\research.json, beside the recent list and
/// the collections. A saved item is the result as the page made it — the question, what was asked of, the
/// ranked evidence and its provenance — kept as it stands and given back unchanged: opening one never runs the
/// research again. The host stores the result whole and only reads the names and the document paths it is told,
/// so the shape of a result stays the page's business.
///
/// No PDF is read, copied or changed: a saved item lists paths, and a file that is gone stays listed (shown as
/// missing) until the item is deleted.
/// </summary>
public sealed class SavedResearch
{
    public const int MaxNameLength = 80;
    /// <summary>A result bigger than this is refused: a saved result is one ranked list, not a document copy.</summary>
    public const int MaxResultLength = 512 * 1024;
    public const int MaxItems = 500;

    private static readonly JsonSerializerOptions Json = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        WriteIndented = true,
    };

    private readonly string _file;
    private readonly object _gate = new();
    private readonly List<SavedResearchItem> _items;

    public SavedResearch(string folder)
    {
        _file = Path.Combine(folder, "research.json");
        _items = Load();
    }

    /// <summary>A copy of every saved research, newest first.</summary>
    public IReadOnlyList<SavedResearchItem> All
    {
        get
        {
            lock (_gate)
            {
                return _items
                    .OrderByDescending(i => i.CreatedAt)
                    .Select(i => new SavedResearchItem { Id = i.Id, Name = i.Name, CreatedAt = i.CreatedAt, Paths = [.. i.Paths], Result = i.Result?.DeepClone() })
                    .ToList();
            }
        }
    }

    /// <summary>True if any saved research quotes this file — what lets the page reopen it at a page.</summary>
    public bool Contains(string path)
    {
        lock (_gate) return _items.Any(i => i.Paths.Any(p => Same(p, path)));
    }

    /// <summary>
    /// Keeps one result under a name. `result` is stored as it is given; `paths` are the documents it quotes,
    /// so a missing file can be shown as missing without looking inside the result.
    /// </summary>
    public SavedResearchItem Save(string name, IEnumerable<string> paths, JsonNode? result)
    {
        var clean = CleanName(name);
        if (result is null) throw new ArgumentException("There is no research result to save.");
        if (result.ToJsonString().Length > MaxResultLength) throw new ArgumentException("That research result is too large to save.");
        var files = (paths ?? []).Where(p => !string.IsNullOrWhiteSpace(p)).Select(Full).Distinct(StringComparer.OrdinalIgnoreCase).ToList();
        SavedResearchItem saved;
        lock (_gate)
        {
            if (_items.Count >= MaxItems) throw new InvalidOperationException($"Vellum keeps at most {MaxItems} saved research results. Delete one first.");
            saved = new SavedResearchItem { Id = Guid.NewGuid().ToString("N"), Name = clean, CreatedAt = DateTimeOffset.Now, Paths = files, Result = result.DeepClone() };
            _items.Add(saved);
        }
        Save();
        return saved;
    }

    /// <summary>Removes that saved research only. Nothing else is touched — no document, no collection, no other item.</summary>
    public void Delete(string id)
    {
        lock (_gate)
        {
            var item = _items.FirstOrDefault(i => i.Id == id) ?? throw new InvalidOperationException("That saved research no longer exists.");
            _items.Remove(item);
        }
        Save();
    }

    private static string Full(string path)
    {
        try { return Path.GetFullPath(path); } catch (Exception) { return path; }
    }

    private static string CleanName(string name)
    {
        var clean = string.Join(' ', (name ?? "").Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
        if (clean.Length == 0) throw new ArgumentException("Saved research needs a name.");
        if (clean.Length > MaxNameLength) throw new ArgumentException($"A name can be at most {MaxNameLength} characters.");
        return clean;
    }

    private static bool Same(string a, string b) => string.Equals(a, b, StringComparison.OrdinalIgnoreCase);

    private List<SavedResearchItem> Load()
    {
        try
        {
            if (!File.Exists(_file)) return [];
            var loaded = JsonSerializer.Deserialize<List<SavedResearchItem?>>(File.ReadAllText(_file), Json) ?? [];
            // A hand-edited or damaged file: an item without an id, a name or a result is dropped rather than
            // shown as an empty result, and the ones that are whole still open.
            return loaded.OfType<SavedResearchItem>()
                .Where(i => !string.IsNullOrWhiteSpace(i.Id) && !string.IsNullOrWhiteSpace(i.Name) && i.Result is not null)
                .GroupBy(i => i.Id, StringComparer.Ordinal)
                .Select(g => g.First())
                .Select(i =>
                {
                    i.Paths = (i.Paths ?? []).Where(p => !string.IsNullOrWhiteSpace(p)).Distinct(StringComparer.OrdinalIgnoreCase).ToList();
                    return i;
                })
                .ToList();
        }
        catch (Exception)
        {
            // A corrupt file starts over, but is kept beside it so nothing a person saved is lost for good.
            try { File.Copy(_file, _file + ".bad", overwrite: true); } catch (Exception) { }
            return [];
        }
    }

    private void Save()
    {
        string json;
        lock (_gate) json = JsonSerializer.Serialize(_items, Json);
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_file)!);
            var temp = _file + ".tmp";
            File.WriteAllText(temp, json);
            File.Move(temp, _file, overwrite: true);
        }
        catch (IOException) { /* best effort; not worth interrupting reading */ }
        catch (UnauthorizedAccessException) { }
    }
}
